python/pyspark/context.py

```diff
@@ -864,8 +865,21 @@ def union(self, rdds):
         first_jrdd_deserializer = rdds[0]._jrdd_deserializer
         if any(x._jrdd_deserializer != first_jrdd_deserializer for x in rdds):
             rdds = [x._reserialize() for x in rdds]
-        cls = SparkContext._jvm.org.apache.spark.api.java.JavaRDD
-        jrdds = SparkContext._gateway.new_array(cls, len(rdds))
+        gw = SparkContext._gateway
+        jvm = SparkContext._jvm
+        jrdd_cls = jvm.org.apache.spark.api.java.JavaRDD
+        jpair_rdd_cls = jvm.org.apache.spark.api.java.JavaPairRDD
+        jdouble_rdd_cls = jvm.org.apache.spark.api.java.JavaDoubleRDD
+        if is_instance_of(gw, rdds[0]._jrdd, jrdd_cls):
+            cls = jrdd_cls
+        elif is_instance_of(gw, rdds[0]._jrdd, jpair_rdd_cls):
```

Comment:
So if the first one is a JavaPairRDD, but another is a JavaRDD, this will fail I think when the array is created on the JVM side? do we want to check whether _all_ are a pair or double RDD? I haven't thought through that but it occurred to me.

Reply:
It will still work:

```python
>>> rdd1 = sc.parallelize([1,2,3,4,5])
>>> rdd2 = sc.parallelize([6,7,8,9,10])
>>> pairRDD1 = rdd1.zip(rdd2)
>>> sc.union([rdd1, pairRDD1]).collect()
[1, 2, 3, 4, 5, (1, 6), (2, 7), (3, 8), (4, 9), (5, 10)]
>>> sc.union([pairRDD1, rdd1]).collect()
[(1, 6), (2, 7), (3, 8), (4, 9), (5, 10), 1, 2, 3, 4, 5]
```

because it converts all to `JavaRDD` when they are different types existent via `self._reserialize()`.

Reply:
And .. actually I think `JavaDoubleRDD` can't be exposed in PySpark if I am not mistaken. It might be unreachable codes but It's just to restore the previous behaviour as is to be extra safe like https://github.com/apache/spark/pull/28648#discussion_r430825553.